shell/platform/android/io/flutter/plugin/text/ProcessTextPlugin.java

```diff
@@ -0,0 +1,192 @@
+  @NonNull private final PackageManager packageManager;
+  @Nullable private ActivityPluginBinding activityBinding;
+  private Map<String, ResolveInfo> resolveInfosById;
+
+  @NonNull
+  private Map<Integer, MethodChannel.Result> requestsByCode =
+      new HashMap<Integer, MethodChannel.Result>();
+
+  public ProcessTextPlugin(@NonNull ProcessTextChannel processTextChannel) {
+    this.processTextChannel = processTextChannel;
+    this.packageManager = processTextChannel.packageManager;
+
+    processTextChannel.setMethodHandler(this);
+  }
+
+  @Override
+  public Map<String, String> queryTextActions() {
+    if (resolveInfosById == null) {
+      resolveInfosById = new HashMap<String, ResolveInfo>();
+      cacheResolveInfos();
+    }
+    Map<String, String> result = new HashMap<String, String>();
+    for (String id : resolveInfosById.keySet()) {
+      final ResolveInfo info = resolveInfosById.get(id);
+      result.put(id, info.loadLabel(packageManager).toString());
+    }
+    return result;
+  }
+
+  @Override
+  public void processTextAction(
+      @NonNull String id,
+      @NonNull String text,
+      @NonNull boolean readOnly,
+      @NonNull MethodChannel.Result result) {
+    if (activityBinding == null) {
+      result.error("error", "Plugin not bound to an Activity", null);
+      return;
+    }
+
+    if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M) {
+      result.error("error", "Android version not supported", null);
+      return;
+    }
+
+    if (resolveInfosById == null) {
+      result.error("error", "Can not process text actions before calling queryTextActions", null);
+      return;
+    }
+
+    final ResolveInfo info = resolveInfosById.get(id);
+    if (info == null) {
+      result.error("error", "Text processing activity not found", null);
+      return;
+    }
+
+    Integer requestCode = result.hashCode();
+    requestsByCode.put(requestCode, result);
+
+    Intent intent = new Intent();
+    intent.setClassName(info.activityInfo.packageName, info.activityInfo.name);
+    intent.setAction(Intent.ACTION_PROCESS_TEXT);
+    intent.setType("text/plain");
+    intent.putExtra(Intent.EXTRA_PROCESS_TEXT, text);
+    intent.putExtra(Intent.EXTRA_PROCESS_TEXT_READONLY, readOnly);
+
+    // Start the text processing activity. When the activity completes, the onActivityResult
+    // callback
+    // is called.
+    activityBinding.getActivity().startActivityForResult(intent, requestCode);
+  }
+
+  private void cacheResolveInfos() {
+    if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M) {
+      return;
+    }
+
+    Intent intent = new Intent().setAction(Intent.ACTION_PROCESS_TEXT).setType("text/plain");
+
+    List<ResolveInfo> infos;
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
+      infos = packageManager.queryIntentActivities(intent, PackageManager.ResolveInfoFlags.of(0));
+    } else {
+      infos = packageManager.queryIntentActivities(intent, 0);
+    }
+
+    resolveInfosById.clear();
```

Comment:
Is this ever called where `resolveInfosById` wouldn't be an empty array? Wondering if we need this call.